src/vset.h

```diff
@@ -0,0 +1,312 @@
+ *-----------------------------------------------------------------------------
+ * Entry Type Contract
+ *-----------------------------------------------------------------------------
+ *
+ * Users must supply a `volatileEntryType` implementation:
+ *
+ *     typedef struct {
+ *         sds (*entryGetKey)(const void *entry);      // get key
+ *         long long (*getExpiry)(const void *entry);  // get expiry
+ *         int (*expire)(void *db, void *o, void *entry); // trigger expiry
+ *     } volatileEntryType;
+ *
+ *-----------------------------------------------------------------------------
+ * Public API
+ *-----------------------------------------------------------------------------
+ *
+ * Create/Free:
+ *     vset *createVolatileSet(volatileEntryType *type);
+ *     void freeVolatileSet(vset *set);
+ *
+ * Mutation:
+ *     int vsetAddEntry(vset *set, void *entry, long long expiry);
+ *     int vsetRemoveEntry(vset *set, void *entry, long long expiry);
+ *     int vsetUpdateEntry(vset *set, void *old_entry,
+ *                                void *new_entry, long long old_expiry,
+ *                                long long new_expiry);
+ *
+ * Expiry Retrieval:
+ *     void *vsetFirstExpired(vset *set, mstime_t now);
+ *     void *vsetPopExpired(vset *set, mstime_t now);
+ *
+ * Utilities:
+ *     bool vsetIsEmpty(vset *set);
+ *
+ * Iteration:
+ *     void vsetStart(vset *set, vsetIterator *it);
+ *     int vsetNext(vsetIterator *it, void **entryptr);
+ *     void vsetStop(vsetIterator *it);
+ *
+ *-----------------------------------------------------------------------------
+ * Iteration Support
+ *-----------------------------------------------------------------------------
+ *
+ * Iterator structure maintains context across all bucket types:
+ *
+ *     typedef struct vsetIterator {
+ *         raxIterator riter;           // for RAX
+ *         hashtableIterator hiter;     // for HT
+ *         uint32_t viter;              // for VECTOR
+ *         void *vsingle;               // for SINGLE
+ *         vsetBucket *parent_bucket;   // owning bucket
+ *         vsetBucket *bucket;          // active bucket
+ *         void *entry;                 // current entry
+ *         long long bucket_ts;         // for RAX
+ *         int iteration_state;         // internal FSM
+ *     } vsetIterator;
+ * */
+
+#define VOLATILESET_BUCKET_INTERVAL_MAX (1LL << 13LL) // 2^13 = 8192 milliseconds
+#define VOLATILESET_BUCKET_INTERVAL_MIN (1LL << 4LL)  // 2^4 = 16 milliseconds
+
+#define VOLATILESET_VECTOR_BUCKET_MAX_SIZE 127
+
+typedef long long (*vsetGetExpiryFunc)(const void *entry);
+
+// Generic bucket type
+typedef void vsetBucket;
+
+// vset is just a pointer to a bucket
+typedef vsetBucket *vset;
+
+typedef struct vsetIterator {
+    /* for rax bucket */
+    raxIterator riter;
+    /* for hashtable bucket */
+    hashtableIterator hiter;
+    /* for vector bucket */
+    uint32_t viter;
+    /* for single bucket */
+    void *vsingle;
```

Comment:
We could wrap these in an anonymous union, since only one of them is used at a time. It makes the sizeof the iterator struct smaller.

```suggestion
    union {
        /* for rax bucket */
        raxIterator riter;
        /* for hashtable bucket */
        hashtableIterator hiter;
        /* for vector bucket */
        uint32_t viter;
        /* for single bucket */
        void *vsingle;
    };
```

Reply:
> We could wrap these in an anonymous union, since only one of them is used at a time. It makes the sizeof the iterator struct smaller.

Yes. although the raxIterator can live while there are other iterators being used. this is why we currently do NOT support nested rax buckets (as we will have to keep track of "stack" while iterating) as I thought it would be very complex for something we have no need to support ATM.
I can try unionizing the rest.

Reply:
> the raxIterator can live while there are other iterators being used

Ah, right. Well, it's easy to unionize some of them later without changing any code accessing them.

Reply:
done for the rest